### PR TITLE
Add documentation on trait checking and type construction

### PR DESCRIPTION
**Pull request by m-rph**
Changes:

Added documentation on how to create types via `Ty` and how to check specific trait implementations.

changelog: none

r? @blyxyas

### CODE REVIEW

File: book/src/development/trait_checking.md [hunk 1]
```diff
@@ -94,6 +94,33 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
 }
 ```
 
+## Creating traits programmatically
```
Comment (flip1995) [inline book/src/development/trait_checking.md:97]: Should this rather be

```suggestion
## Creating Types Programmatically
```

Definitely should be Title Case.

File: book/src/development/trait_checking.md [hunk 2]
```diff
@@ -94,6 +94,33 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
+
+Traits are often generic over a type e.g. `Borrow<T>` is generic over `T`, and rust allows us to implement a trait for
+a specific type. For example, we can implement `Borrow<str>` for a hypothetical type `Foo`. Let's suppose that we
```
Comment (flip1995) [inline book/src/development/trait_checking.md:100]:
```suggestion
a specific type. For example, we can implement `Borrow<[u8]>` for a hypothetical type `Foo`. Let's suppose that we
```

As you use `[u8]` in the rest of the example, I would use that throughout.

File: book/src/development/trait_checking.md [hunk 3]
```diff
@@ -94,6 +94,33 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
+would like to find whether our type actually implements `Borrow<[u8]>`. To do so, we need to supply a type that
+represents `[u8]`, but `[u8]` is also a generic, it's a slice over `u8`. We can create this type using Ty::new_slice
```
Comment (flip1995) [inline book/src/development/trait_checking.md:102]:
```suggestion
represents `[u8]`, but `[u8]` is also a generic, it's a slice over `u8`. We can create this type using [`Ty::new_slice`][new_slice]
```

And then link

```md
[new_slice]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.new_slice
```

below.

File: book/src/development/trait_checking.md [hunk 4]
```diff
@@ -94,6 +94,33 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
+method. The following code demonstrates how to do this:
+
+```rust
+
+use rustc_middle::ty::Ty;
+use clippy_utils::ty::implements_trait;
+use rustc_span::symbol::sym;
+
+let ty = todo!();
```
Comment (flip1995) [inline book/src/development/trait_checking.md:111]: Please add a str of what the `todo` is about:

```suggestion
let ty = todo!("Get type `Foo` to check for a trait implementation");
```

File: book/src/development/trait_checking.md [hunk 5]
```diff
@@ -94,6 +94,33 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
+let borrow_id = cx.tcx.get_diagnostic_item(sym::Borrow).unwrap(); // avoid unwrap in real code
+if implements_trait(cx, ty, borrow_id, &[Ty::new_slice(cx.tcx, cx.tcx.types.u8).into()]) {
+    todo!()
```
Comment (flip1995) [inline book/src/development/trait_checking.md:114]:
```suggestion
    todo!("Rest of lint implementation");
```

File: book/src/development/trait_checking.md [hunk 6]
```diff
@@ -94,6 +94,33 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
+
+```rust
+
+use rustc_middle::ty::Ty;
+use clippy_utils::ty::implements_trait;
+use rustc_span::symbol::sym;
+
+let ty = todo!();
+let borrow_id = cx.tcx.get_diagnostic_item(sym::Borrow).unwrap(); // avoid unwrap in real code
+if implements_trait(cx, ty, borrow_id, &[Ty::new_slice(cx.tcx, cx.tcx.types.u8).into()]) {
```
Comment (flip1995) [inline book/src/development/trait_checking.md:113]: I would place the `Ty::new_slice` in its own line before the `if`, as this is what this documentation is about. Then maybe add a comment like

```rust
let slice = Ty::new_slice(...) // <- creating the type here
```

**Review by flip1995 — CHANGES_REQUESTED** [review]
Thanks! A lot of nit picking.

**Comment by PartiallyTyped** [comment]
Thanks! I will correct later today ☺️

File: book/src/development/trait_checking.md [hunk 7]
```diff
@@ -96,29 +96,36 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
-would like to find whether our type actually implements `Borrow<[u8]>`. To do so, we need to supply a type that
-represents `[u8]`, but `[u8]` is also a generic, it's a slice over `u8`. We can create this type using Ty::new_slice
-method. The following code demonstrates how to do this:
+Traits are often generic over a type parameter, e.g. `Borrow<T>` is generic
+over `T`. Rust allows us to implement a trait for a specific type. For example,
+we can implement `Borrow<[u8]>` for a hypothetical type `Foo`. Let's suppose
+that we would like to find whether our type actually implements `Borrow<[u8]>`.
+
+To do so, we can use the same `implements_trait` function as above, and supply
+a parameter type that represents `[u8]`. Since `[u8]` is a specialization of
```
Comment (PartiallyTyped) [inline book/src/development/trait_checking.md:105]: Should this be `parameter type` or `type parameter`?

File: book/src/development/trait_checking.md [hunk 8]
```diff
@@ -96,29 +96,36 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
+`[T]`, we can use the  [`Ty::new_slice`][new_slice] method to create a type
+that represents `[T]` and supply `u8` as a type parameter. The following code
```
Comment (PartiallyTyped) [inline book/src/development/trait_checking.md:107]: Should we mention that we have to turn this value into a generic first?

File: book/src/development/trait_checking.md [hunk 9]
```diff
@@ -129,5 +136,6 @@ useful when we need to create types that we can't obtain through the usual means
 [symbol]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/symbol/struct.Symbol.html
 [symbol_index]: https://doc.rust-lang.org/beta/nightly-rustc/rustc_span/symbol/sym/index.html
 [TyCtxt]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html
-[Ty]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html
+[`Ty`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html
```
Comment (PartiallyTyped) [inline book/src/development/trait_checking.md:139]: This should be rolled back.

**Review by PartiallyTyped — COMMENTED** [review]
I think this looks slightly better, but need to fix that `Ty` link.

File: book/src/development/trait_checking.md (hunk 7 above)
Comment (flip1995) [reply book/src/development/trait_checking.md:105]: I would say type parameter :thinking:

File: book/src/development/trait_checking.md (hunk 8 above)
Comment (flip1995) [reply book/src/development/trait_checking.md:107]: I would add a `// <- turn into generic parameter` comment in the code below. Also point and comment the `new_slice` line in the code below.

File: book/src/development/type_checking.md [hunk 10]
```diff
@@ -123,6 +123,26 @@ the [`TypeckResults::node_type()`][node_type] method inside of bodies.
 
 > **Warning**: Don't use `hir_ty_to_ty` inside of bodies, because this can cause ICEs.
 
+## Creating Types programmatically
+
+A common usecase for creating types programmatically is when we want to check if a type implements a trait. We have
+a section on this in the [Trait Checking](trait_checking.md) chapter, but given the importance of this topic, we will
+also cover it a bit here.
```
[the remark this answers is not in the document]
Comment (PartiallyTyped) [reply book/src/development/type_checking.md:130]: That makes sense!

[the remark this answers is not in the document]
Comment (PartiallyTyped) [reply book/src/development/type_checking.md:130]: Speaking of, do you have any other patterns off the top of your head that should be documented?

[the remark this answers is not in the document]
Comment (flip1995) [reply book/src/development/type_checking.md:130]: Not really. I sadly don't have the time to contribute actual code to Clippy anymore :smiling_face_with_tear:  That's why I'm a bit OOTL when it comes to patterns for lint implementations.

File: book/src/development/trait_checking.md [hunk 11]
```diff
@@ -94,6 +94,53 @@ impl LateLintPass<'_> for CheckTokioAsyncReadExtTrait {
...
+
+To do so, we can use the same `implements_trait` function as above, and supply
+a type parameter that represents `[u8]`. Since `[u8]` is a specialization of
+`[T]`, we can use the  [`Ty::new_slice`][new_slice] method to create a type
+that represents `[T]` and supply `u8` as a type parameter.
+To create a `ty::Ty` programmatically, we rely on `Ty::new_*` methods. These
+methods create a `TyKind` and then wrap it in a `Ty` struct. This means we
+have access to all the primitive types, such as `Ty::new_char`,
+`Ty::new_bool`, `Ty::new_int`, etc. We can also create more complex types,
+such as slices, tuples, and references out of these basic building blocks.
```
Comment (flip1995) [inline book/src/development/trait_checking.md:112]: Fun fact: [`slice`](https://doc.rust-lang.org/std/primitive.slice.html) is actually a primitive type. But I think not listing it here is fine and better this way.

**Review by flip1995 — APPROVED** [review]
Thanks!

**Pull request merged**